### PR TITLE
Keys and Cards match contents (for real)

### PR DESCRIPTION
**Pull request by GameWyrm**
This makes the "Chests Match Contents" setting also apply to freestanding keys and cards. You can now preview these items to see what they are, and in the case of keys the appearance will hint you on what dungeon it's for. Traps and remote items are supported.

Remake of https://github.com/Extra-2-Dew/ArchipelagoRandomizer/pull/15 to the correct branch

### CODE REVIEW

File: Code/FreestandingReplacer.cs [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+        }
+
+        /// <summary>
+        /// Finds an object to use for previewing from the given path. Automatically adjusts for ItemSelectors.
+        /// </summary>
+        /// <param name="itemName">The item's key in previewItemData.json</param>
+        /// <returns></returns>
+        public static GameObject GetModelFromPath(string itemName)
+        {
+            Plugin.Log.LogInfo($"Preloading {itemName}...");
```
Comment (Chris-Is-Awesome) [inline Code/FreestandingReplacer.cs:79]: Remove testing logs when finished testing so we don't clutter it.

**Review by Chris-Is-Awesome — CHANGES_REQUESTED** [review]
Some small changes would be good, but overall looks good

File: Code/FreestandingReplacer.cs [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+            {
+                model = GameObject.Instantiate(GetGameObjectFromSelector(observer, data.index));
+            }
+
+            return GenerateModelPrefab(model, data, itemName);
+        }
+
+        public static GameObject GetModelFromDroptable(string itemName)
+        {
+            Plugin.Log.LogInfo($"Preloading {itemName}...");
```
[the remark this answers is not in the document]
Comment (GameWyrm) [reply Code/FreestandingReplacer.cs]: Removed for now.

File: Code/FreestandingReplacer.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (GameWyrm) [reply Code/FreestandingReplacer.cs]: Removed for now

File: Code/Patches.cs [hunk 3]
```diff
@@ -127,6 +128,20 @@ public static bool MainMenu_DoStartMenu_Patch(MainMenu __instance)
 			return false;
 		}
 
+		// Replace key with model of randomized item
+		[HarmonyPostfix]
+		[HarmonyPatch(typeof(SpawnItemEventObserver), nameof(SpawnItemEventObserver.SpawnItem))]
+		public static void SpawnItemEventObserver_SpawnItem_Postfix(SpawnItemEventObserver __instance)
+		{
+			if (!ItemRandomizer.IsActive) return;
```
[the remark this answers is not in the document]
Comment (GameWyrm) [reply Code/Patches.cs]: Fixed

**Pull request merged**